### PR TITLE
Backend refactorings and perf improvements

### PR DESCRIPTION
**Pull request by alexcrichton**
This PR is extracted from https://github.com/rust-lang/cargo/pull/6864 and then also additionally adds some commits related to performance optimizations that I noticed while profiling https://github.com/rust-lang/cargo/pull/6864. Each commit is in theory standalone and should pass all the tests, as well as being descriptive about what it's doing.

### CODE REVIEW

**Comment by alexcrichton** [comment]
I should mention that the overall performance wins of this were pretty modest, I was seeing ~10% with an average of 180-190ms for a null build of Cargo itself, whereas afterwards it's 170-180ms but the variances are quite high.

**Comment by alexcrichton** [comment]
I should also note that the lion's share of a possible improvement to Cargo null build time will likely be https://github.com/rust-lang/cargo/issues/6866 as that's >50% of a null build from my measurements. While we can't get it down to zero probably we can make it way faster in theory.

File: src/cargo/core/compiler/unit.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            kind,
+            mode,
+        });
+        Unit { inner }
+    }
+
+    fn intern_inner(&'a self, item: &UnitInner<'a>) -> &'a UnitInner<'a> {
+        let mut me = self.state.borrow_mut();
+        if let Some(item) = me.cache.get(item) {
+            return unsafe { &*(&**item as *const UnitInner<'a>) };
```
Comment (ehuss) [inline src/cargo/core/compiler/unit.rs]: I think I understand what this is doing, but can you add a comment explaining it, along with why it is safe?

Is this correct: Starting with `&Box<UnitInner>` it pulls `UnitInner` out of the box, casts to a pointer to force the borrow checker to forget about the borrow, then converts the pointer back to `&UnitInner`?

Comment (alexcrichton) [reply src/cargo/core/compiler/unit.rs]: Correct yeah, and I'll leave some comments, an excellent point!

**Comment by ehuss** [comment]
Hmm, something looks wrong with rustup, it's failing for me elsewhere.

**Pull request merged**